### PR TITLE
Proof-of-concept change that ignores exceptions from subprocesses.

### PR DESCRIPTION
**Pull request by vorporeal**
## :pencil: Checklist

- [ ] I reviewed the submitted code.
- [ ] I added tests to verify the changes.
- [ ] No new PII added or SDK only sends newly added PII if `sendDefaultPII` is enabled.
- [ ] I updated the docs if needed.
- [ ] Review from the native team if needed.
- [ ] No breaking change or entry added to the changelog.
- [ ] No breaking change for hybrid SDKs or communicated to hybrid SDKs.

### CODE REVIEW

File: Sources/SentryCrash/Recording/Monitors/SentryCrashMonitor_MachException.c [hunk 1]
```diff
@@ -313,7 +313,30 @@ handleExceptions(void *const userData)
...
             sizeof(exceptionMessage), g_exceptionPort, MACH_MSG_TIMEOUT_NONE, MACH_PORT_NULL);
         if (kr == KERN_SUCCESS) {
-            break;
+            // If the exception was generated in a different process than the one in which
+            // the handler was installed, ignore it.  Task-level exception handlers are
+            // inherited by subprocesses, and we don't want to handle or report exceptions
+            // in arbitrary subprocesses of the one in which we were installed.
+            if (exceptionMessage.task.name != mach_task_self()) {
+                SentryCrashLOG_DEBUG("Received exception from subprocess. Ignoring message and waiting for another.");
+                // Send a reply saying "I didn't handle this exception".
```
Comment (vorporeal) [inline Sources/SentryCrash/Recording/Monitors/SentryCrashMonitor_MachException.c:322]: There is similar code later in this function which claims to reply and say "I didn't handle this exception".  I initially tried using that code here, but it did not appear to work correctly - subprocesses would not terminate post-exception as one would have expected.

Comment (philipphofmann) [reply Sources/SentryCrash/Recording/Monitors/SentryCrashMonitor_MachException.c:322]: Thanks a lot for opening up this draft PR, @vorporeal 🚀. At a first glance, this PR makes sense. @armcknight I think you have more knowledge on mach messages. Could you please have a look?

File: Sources/SentryCrash/Recording/Monitors/SentryCrashMonitor_MachException.c [hunk 2]
```diff
@@ -313,7 +313,30 @@ handleExceptions(void *const userData)
...
+                replyMessage.header.msgh_id = 0;
```
Comment (philipphofmann) [inline Sources/SentryCrash/Recording/Monitors/SentryCrashMonitor_MachException.c:323]: `m`: I'm unsure if we can always set this to 0. I guess it's better to reference the existing id and create a new one with something like. WDYT, @armcknight?
```suggestion
                replyMessage.header.msgh_id = exceptionMessage.header.msgh_id + 1000
```

File: Sources/SentryCrash/Recording/Monitors/SentryCrashMonitor_MachException.c [hunk 3]
```diff
@@ -313,7 +313,30 @@ handleExceptions(void *const userData)
...
+                replyMessage.header.msgh_bits = exceptionMessage.header.msgh_bits & MACH_MSGH_BITS_REMOTE_MASK;
+                replyMessage.header.msgh_remote_port = exceptionMessage.header.msgh_remote_port;
+                replyMessage.header.msgh_local_port = MACH_PORT_NULL;
+                replyMessage.header.msgh_size = sizeof(mach_msg_header_t);
```
Comment (philipphofmann) [inline Sources/SentryCrash/Recording/Monitors/SentryCrashMonitor_MachException.c:327]: `m`: I think this should be the reply message size and not the header.
```suggestion
                replyMessage.header.msgh_size = sizeof(MachReplyMessage);
```

File: Sources/SentryCrash/Recording/Monitors/SentryCrashMonitor_MachException.c [hunk 4]
```diff
@@ -313,7 +313,30 @@ handleExceptions(void *const userData)
...
+
+                replyMessage.NDR = exceptionMessage.NDR;
+                // Tell the kernel that we failed to handle the exception, so that
+                // the next handler in the chain gets invoked.
+                replyMessage.returnCode = KERN_FAILURE;
+
+                mach_msg(&replyMessage.header, MACH_SEND_MSG, sizeof(replyMessage), 0, replyMessage.header.msgh_remote_port,
```
Comment (philipphofmann) [inline Sources/SentryCrash/Recording/Monitors/SentryCrashMonitor_MachException.c:334]: `m`: Why did you choose `replyMessage.header.msgh_remote_port` and not `MACH_PORT_NULL`, like we do in https://github.com/getsentry/sentry-cocoa/blob/0454d35faead8e753690921f088809b68c2d763d/Sources/SentryCrash/Recording/Monitors/SentryCrashMonitor_MachException.c#L401-L402

**Review by philipphofmann — COMMENTED** [review]
Maybe a bit unrelated or maybe not: @armcknight, shouldn't we send `KERN_SUCESS` here 

https://github.com/getsentry/sentry-cocoa/blob/0454d35faead8e753690921f088809b68c2d763d/Sources/SentryCrash/Recording/Monitors/SentryCrashMonitor_MachException.c#L399-L402

File: Sources/SentryCrash/Recording/Monitors/SentryCrashMonitor_MachException.c (hunk 4 above)
Comment (vorporeal) [reply Sources/SentryCrash/Recording/Monitors/SentryCrashMonitor_MachException.c]: I think I may have misunderstood the meaning of that parameter (I parsed it as the destination port for the message being sent) - looks like, from additional testing, that the key line is setting `replyMessage.header.msgh_remote_port`, and that passing `MACH_PORT_NULL` here is fine.

Updated the PR accordingly.

**Comment by vorporeal** [comment]
> Maybe a bit unrelated or maybe not: @armcknight, shouldn't we send `KERN_SUCESS` here
> 
> https://github.com/getsentry/sentry-cocoa/blob/0454d35faead8e753690921f088809b68c2d763d/Sources/SentryCrash/Recording/Monitors/SentryCrashMonitor_MachException.c#L399-L402

The code you linked might need to be updated along the lines of what I'm doing in this PR.  My understanding is that, if you send `KERN_SUCCESS`, you're telling the kernel that the exception has actually been _handled_ and that it can continue executing the program that produced the exception.

**Review by armcknight — COMMENTED** [review]
Thanks for the PR @vorporeal! Very interesting, haven't thought about it before (as most of my experience is from iOS, where subprocesses aren't something to worry about). 

How did you validate this? I tried updating our macOS-Swift test app to add a new button that will launch a subprocess, a new CLI tool target I added to the same project that crashes with a mach_exception when it launches. But I'm having a bit of trouble validating that there is new behavior, maybe @philipphofmann can help me see where some output of the mach exception handler could be found that we would expect to be absent with the new changes. 

You can see the test harness I started building in https://github.com/vorporeal/sentry-cocoa/pull/1

File: Sources/SentryCrash/Recording/Monitors/SentryCrashMonitor_MachException.c (hunk 2 above)
Comment (armcknight) [reply Sources/SentryCrash/Recording/Monitors/SentryCrashMonitor_MachException.c:323]: see also https://www.gnu.org/software/hurd/gnumach-doc/Message-Format.html:
> The mach_msg call doesn't use the msgh_id field, but it conventionally conveys an operation or function id.

What happens if we keep assign the existing `exceptionMessage.header.msgh_id`?

**Comment by philipphofmann** [comment]
> maybe @philipphofmann can help me see where some output of the mach exception handler could be found that we would expect to be absent with the new changes.

I think the key to validating is that when the subprocess raises a match exception the host process doesn't create the crash report as pointed out here https://github.com/getsentry/sentry-cocoa/issues/2695.

File: Sources/SentryCrash/Recording/Monitors/SentryCrashMonitor_MachException.c (hunk 2 above)
Comment (vorporeal) [reply Sources/SentryCrash/Recording/Monitors/SentryCrashMonitor_MachException.c:323]: Seems to work fine using `exceptionMessage.header.msgh_id`.  Would you prefer that over using 0 with a comment?

**Comment by vorporeal** [comment]
@armcknight Another behavior I noticed before these changes is that the second time you run a subprocess that crashes, it never actually terminates - it stays running, presumably stuck waiting for the Sentry exception handler to respond.

File: Sources/SentryCrash/Recording/Monitors/SentryCrashMonitor_MachException.c [hunk 5]
```diff
@@ -313,7 +313,30 @@ handleExceptions(void *const userData)
         kern_return_t kr = mach_msg(&exceptionMessage.header, MACH_RCV_MSG, 0,
             sizeof(exceptionMessage), g_exceptionPort, MACH_MSG_TIMEOUT_NONE, MACH_PORT_NULL);
         if (kr == KERN_SUCCESS) {
-            break;
+            // If the exception was generated in a different process than the one in which
+            // the handler was installed, ignore it.  Task-level exception handlers are
+            // inherited by subprocesses, and we don't want to handle or report exceptions
+            // in arbitrary subprocesses of the one in which we were installed.
+            if (exceptionMessage.task.name != mach_task_self()) {
```
[the remark this answers is not in the document]
Comment (vorporeal) [reply Sources/SentryCrash/Recording/Monitors/SentryCrashMonitor_MachException.c]: Totally agree; done.  (I wasn't sure if this was going to get merged "directly", so I didn't do a thorough code cleanup pass before linking it in the bug I filed.)

File: Sources/SentryCrash/Recording/Monitors/SentryCrashMonitor_MachException.c (hunk 2 above)
Comment (armcknight) [reply Sources/SentryCrash/Recording/Monitors/SentryCrashMonitor_MachException.c]: I think it makes sense to use the preexisting `msgh_id` 👍🏻

**Comment by armcknight** [comment]
> @armcknight Another behavior I noticed before these changes is that the second time you run a subprocess that crashes, it never actually terminates - it stays running, presumably stuck waiting for the Sentry exception handler to respond.

I didn't observe that in the test harness I set up, but it would be good for either you or @philipphofmann to give it a try in case it's not correctly set up. I see the same termination status multiple times in the same process in my console:

```
finished; status: 11; reason: NSTaskTerminationReason(rawValue: 2)
finished; status: 11; reason: NSTaskTerminationReason(rawValue: 2)
finished; status: 11; reason: NSTaskTerminationReason(rawValue: 2)
```

This is from just pressing the new "Crash in subprocess" button at the bottom over and over:
image

**Comment by vorporeal** [comment]
I'm trying out your harness, and I'm not sure whether it's due to the small configuration changes I needed to make to get it compile or some other reason, but it's not clear to me that the crash integration/mach exception handler is actually getting set up properly.  When I run the test harness then pause execution and look at the list of threads, I don't see either the primary or secondary crash reporting handler threads ("SentryCrash Exception Handler (Primary)" and "SentryCrash Exception Handler (Secondary)".

Are you sure the integration is getting properly initialized?

**Comment by armcknight** [comment]
> I'm trying out your harness, and I'm not sure whether it's due to the small configuration changes I needed to make to get it compile or some other reason, but it's not clear to me that the crash integration/mach exception handler is actually getting set up properly. When I run the test harness then pause execution and look at the list of threads, I don't see either the primary or secondary crash reporting handler threads ("SentryCrash Exception Handler (Primary)" and "SentryCrash Exception Handler (Secondary)".
> 
> Are you sure the integration is getting properly initialized?

If you have a debugger attached, the mach exception handler will not be installed. See https://github.com/getsentry/sentry-cocoa/blob/63f740dc5141fc10b864f36e13eedd3d8bff79dd/Sources/SentryCrash/Recording/Monitors/SentryCrashMonitorType.h#L63-L83

Unfortunately if you want to debug here, it's going to be a pain. You have to resort to things like writing log statements out to a file.

**Comment by vorporeal** [comment]
Ah, that would do it, thanks!  Running it manually (outside of XCode) produces certain expected behaviors before and after the change:

Before:
* Crashing the subprocess triggers the mach exception handler
* Subsequently crashing the main process (via the `Sentry.crash()` button) does _not_ trigger the mach exception handler

After:
* Crashing the subprocess prints out the following debug message (which I added): "Received exception from subprocess. Ignoring message and waiting for another."
* Subsequently crashing the main process does trigger the mach exception handler

I'm not sure why the subprocesses fail to terminate in the "before" case in my testing with our app (Warp) but not in the test harness.  Nevertheless, the behavioral change described above is itself an important issue worth addressing IMO.

File: Sources/SentryCrash/Recording/Monitors/SentryCrashMonitor_MachException.c (hunk 2 above)
Comment (vorporeal) [reply Sources/SentryCrash/Recording/Monitors/SentryCrashMonitor_MachException.c]: Done.

**Comment by philipphofmann** [comment]
> > Maybe a bit unrelated or maybe not: @armcknight, shouldn't we send `KERN_SUCESS` here
> > https://github.com/getsentry/sentry-cocoa/blob/0454d35faead8e753690921f088809b68c2d763d/Sources/SentryCrash/Recording/Monitors/SentryCrashMonitor_MachException.c#L399-L402
> 
> The code you linked might need to be updated along the lines of what I'm doing in this PR. My understanding is that, if you send `KERN_SUCCESS`, you're telling the kernel that the exception has actually been _handled_ and that it can continue executing the program that produced the exception.
> 
> See: https://gist.github.com/rodionovd/01fff61927a665d78ecf#file-mach_exceptions-cpp-L240-L262
> 
> That all said, when I change my newly-added logic to return `KERN_SUCCESS` instead of `KERN_FAILURE`, the desired behavior (the subprocess terminates without generating a crash report) is maintained. Not sure whether returning `KERN_SUCCESS` would prevent escalation up to the Apple crash reporter, though; I don't have an easy way to test that. (I wish Apple had better documentation.)

I would stick to `KERN_FAILURE` because it allows other mach exception handlers to react to the exception. We have used it for years and it's working fine. If you use our SDK alongside other crash reporting libraries `KERN_FAILURE` allows them also to create a crash report. Using `KERN_SUCCESS` could prevent that.

**Comment by philipphofmann** [comment]
> Nevertheless, the behavioral change described above is itself an important issue worth addressing IMO.

What exactly do you mean by that @vorporeal?

**Comment by philipphofmann** [comment]
If I'm correct, we addressed all the issues in this PR, @armcknight, @vorporeal? Is there anything missing?

The following steps are to ensure we are not breaking existing functionality and to somehow test this new behavior in CI. I think the first step would be to include https://github.com/vorporeal/sentry-cocoa/pull/1 in this PR here to have at least a manual way to test this. @vorporeal, do you have ideas on how to test this in CI, or should I take the lead from here on adding them?

**Comment by vorporeal** [comment]
> > Nevertheless, the behavioral change described above is itself an important issue worth addressing IMO.
> 
> What exactly do you mean by that @vorporeal?

Was only saying that even if the "crashed subprocesses never fully terminate" issue doesn't reproduce in all contexts (i.e.: it does in our Sentry-embedding application but doesn't in the manual test harness), this change is still important in order to ensure that a crash in a subprocess doesn't prevent reporting of a subsequent crash in the main process.

@philipphofmann In terms of getting automated tests on CI - I think it makes more sense for you to take that on.  I don't have enough familiarity with your codebase (or macOS development in general) to be able to effectively or efficiently contribute there.

I've merged @armcknight's manual test harness change into this PR.

**Comment by philipphofmann** [comment]
Thanks for the clarification, @vorporeal. I will devise an automated way to test this, but I can't give you an ETA yet.

**Comment by armcknight** [comment]
This is a nice piece of context that didn't really come across to me the last couple times I looked:

> this change is still important in order to ensure that a crash in a subprocess doesn't prevent reporting of a subsequent crash in the main process

I think we have some decisions to make in terms of technicals and product direction here. As a developer thinking like a Sentry customer, I might want to know about crashes in my macOS app's subprocesses, of course in addition to any subsequent crashes in the macOS app itself. Maybe subprocess crashes would be reported as errors to Sentry for the main macOS app. That could be helpful information to have if a crash in a subprocess ultimately led to a crash or error in the main app.

**Comment by armcknight** [comment]
This PR hasn't seen any progress in a bit, and I think we're still unclear on the tradeoffs of this approach. We need a reproduction so we can observe the desired behavior, as well as to use in a test case going forward. We'll close this for now, but please reopen or start a new one if more can be done. We will also leave #2695 open for anyone that is interested in moving it forward. Thanks again for your contributions on bringing up the issue and laying the groundwork for the potential changes @vorporeal.

**Pull request closed without merging**